File: typescript/packages/langchain/src/agent.ts
import { createReactAgent } from '@langchain/langgraph/prebuilt';
import type { BaseChatModel } from '@langchain/core/language_models/chat_models';
import type { StructuredToolInterface } from '@langchain/core/tools';
import type { Runnable, RunnableConfig } from '@langchain/core/runnables';
import type { BaseMessage } from '@langchain/core/messages';
import type { CopassClient, ContextWindow, SearchPreset } from '@copass/core';
import { copassTools } from './tools.js';
import { CopassWindowCallback } from './callback.js';

type ReactAgentExtraOptions = Omit<
  Parameters<typeof createReactAgent>[0],
  'llm' | 'tools'
>;

/**
 * Output type of {@link createCopassAgent}. A standard LangChain `Runnable`
 * producing `{ messages: BaseMessage[] }` on `.invoke()` / `.stream()`.
 *
 * The concrete inferred type (a LangGraph `CompiledStateGraph`) can't be
 * serialized into a portable declaration, so we expose the narrower
 * `Runnable` surface that's sufficient for every public method a caller
 * needs — `.invoke`, `.stream`, `.streamEvents`, `.batch`, `.pipe`,
 * `.withConfig`, etc. Escape hatch if you need the raw graph: cast the
 * result.
 */
export type CopassAgent = Runnable<
  { messages: BaseMessage[] } | BaseMessage[] | string,
  { messages: BaseMessage[] },
  RunnableConfig
>;

export interface CreateCopassAgentOptions extends ReactAgentExtraOptions {
  /** An authenticated `@copass/core` client. */
  client: CopassClient;
  /** Sandbox all retrieval runs against. */
  sandbox_id: string;
  /**
   * Context Window for this conversation thread. Turn history is mirrored
   * into it automatically across every chat model invocation — retrieval
   * becomes window-aware with no per-call wiring.
   */
  window: ContextWindow;
  /** The chat model (e.g. `new ChatAnthropic(...)`, `new ChatOpenAI(...)`). */
  llm: BaseChatModel;
  /** Additional tools to mix in alongside `discover` / `interpret` / `search`. */
  tools?: StructuredToolInterface[];
  /** Optional project scoping for retrieval. */
  project_id?: string;
  /** Preset for `interpret` / `search`. Default: `"fast"`. */
  preset?: SearchPreset;
  /** Include LangChain `ToolMessage`s in the Context Window. Default: `false`. */
  includeToolMessages?: boolean;
}

/**
 * Create a LangChain agent pre-wired with Copass retrieval tools and
 * window-aware memory.
 *
 * Returns a standard LangChain Runnable — call `.invoke()`, `.stream()`,
 * `.streamEvents()`, `.batch()` as you would with the output of
 * `createReactAgent`. The Copass Context Window is mirrored from the
 * conversation history automatically; you don't pass callbacks, trackers,
 * or lifecycle hooks.
 *
 * @example
 * ```ts
 * import { ChatAnthropic } from '@langchain/anthropic';
 * import { CopassClient } from '@copass/core';
 * import { createCopassAgent } from '@copass/langchain';
 *
 * const copass = new CopassClient({ auth: { type: 'bearer', token: process.env.COPASS_API_KEY! } });
 * const window = await copass.contextWindow.create({ sandbox_id });
 *
 * const agent = createCopassAgent({
 *   client: copass,
 *   sandbox_id,
 *   window,
 *   llm: new ChatAnthropic({ model: 'claude-opus-4-7' }),
 * });
 *
 * const result = await agent.invoke({
 *   messages: [{ role: 'user', content: 'why is checkout flaky?' }],
 * });
 * ```
 *
 * ### How the wiring works
 *
 * Under the hood this composes three primitives also exported from this
 * package — in case you want to build your own variant:
 *
 * - `copassTools({ client, sandbox_id, window, ... })` — the three retrieval
 *   tools, window-aware at the server call level.
 * - `createReactAgent({ llm, tools, ... })` — the standard LangGraph ReAct agent.
 * - `CopassWindowCallback({ window })` — bound via `Runnable.withConfig` so
 *   every chat model invocation auto-mirrors messages into the window.
 */
export function createCopassAgent(options: CreateCopassAgentOptions): CopassAgent {
  const {
    client,
    sandbox_id,
    window,
    llm,
    tools: extraTools = [],
    project_id,
    preset,
    includeToolMessages,
    ...reactAgentOptions
  } = options;

  const copass = copassTools({ client, sandbox_id, window, project_id, preset });
  const callback = new CopassWindowCallback({ window, includeToolMessages });

  const agent = createReactAgent({
    llm,
    tools: [copass.discover, copass.interpret, copass.search, ...extraTools],
    ...reactAgentOptions,
  });

  // `withConfig` returns a RunnableBinding that propagates these callbacks
  // to every child Runnable invocation — including the chat model calls
  // that trigger `CopassWindowCallback.handleChatModelStart`. The cast
  // narrows LangGraph's concrete graph type to the portable Runnable
  // surface we want to expose.
  return agent.withConfig({ callbacks: [callback] }) as unknown as CopassAgent;
}

File: python/copass-langchain/src/copass_langchain/agent.py
"""Convenience factory for a Copass-wired LangGraph ReAct agent.

Python mirror of ``typescript/packages/langchain/src/agent.ts`` — wires
:func:`copass_tools` + :class:`CopassWindowCallback` (when a window is
supplied) into ``langgraph.prebuilt.create_react_agent``. Returns the
agent with the callback pre-bound via ``agent.with_config(...)``.

``langgraph`` is an **optional** dependency — install with
``pip install copass-langchain[agent]`` to use :func:`create_copass_agent`.
The core :func:`copass_tools` + :class:`CopassWindowCallback` exports
work without it.
"""

from __future__ import annotations

from typing import Any, List, Optional

from copass_core import CopassClient, SearchPreset

from copass_langchain.callback import CopassWindowCallback
from copass_langchain.tools import copass_tools
from copass_langchain.types import ContextWindowLike


def create_copass_agent(
    *,
    client: CopassClient,
    sandbox_id: str,
    llm: Any,
    window: Optional[ContextWindowLike] = None,
    tools: Optional[List[Any]] = None,
    project_id: Optional[str] = None,
    preset: SearchPreset = "copass/copass_1.0",
    include_tool_messages: bool = False,
    **react_agent_options: Any,
) -> Any:
    """Build a LangGraph ReAct agent pre-wired with Copass retrieval.

    Args:
        client: Authenticated :class:`CopassClient`.
        sandbox_id: Sandbox every retrieval call runs against.
        llm: A LangChain chat model instance (``ChatAnthropic(...)``,
            ``ChatOpenAI(...)``, etc.).
        window: Optional :class:`ContextWindowLike`. When supplied, a
            :class:`CopassWindowCallback` is bound so conversation
            history auto-mirrors into the window and retrieval
            becomes window-aware.
        tools: Additional tools to mix in alongside the three Copass
            retrieval tools.
        project_id: Optional project narrowing for retrieval.
        preset: Preset for ``interpret`` / ``search``.
        include_tool_messages: Whether :class:`CopassWindowCallback`
            mirrors tool-result messages. Default ``False``.
        **react_agent_options: Forwarded to
            ``langgraph.prebuilt.create_react_agent`` (e.g.,
            ``checkpointer``, ``state_schema``, ``prompt``).

    Returns:
        A LangChain ``Runnable`` — call ``.ainvoke({"messages": [...]})``.

    Raises:
        ImportError: If ``langgraph`` isn't installed. Install with
            ``pip install copass-langchain[agent]``.

    Example::

        from copass_core import ApiKeyAuth, CopassClient
        from copass_langchain import create_copass_agent
        from langchain_anthropic import ChatAnthropic

        client = CopassClient(auth=ApiKeyAuth(key="olk_..."))
        agent = create_copass_agent(
            client=client,
            sandbox_id="sb_...",
            llm=ChatAnthropic(model="claude-opus-4-7"),
        )
        result = await agent.ainvoke(
            {"messages": [("user", "why is checkout flaky?")]},
        )
    """
    try:
        from langgraph.prebuilt import create_react_agent
    except ImportError as e:
        raise ImportError(
            "create_copass_agent requires `langgraph`. Install with "
            "`pip install copass-langchain[agent]`."
        ) from e

    copass = copass_tools(
        client=client,
        sandbox_id=sandbox_id,
        project_id=project_id,
        window=window,
        preset=preset,
    )

    all_tools: List[Any] = copass.all() + list(tools or [])
    agent = create_react_agent(model=llm, tools=all_tools, **react_agent_options)

    if window is not None:
        callback = CopassWindowCallback(
            window=window,
            include_tool_messages=include_tool_messages,
        )
        return agent.with_config({"callbacks": [callback]})

    return agent


__all__ = ["create_copass_agent"]
